pkg/kubelet/status_manager.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+)
+
+type podStatusSyncRequest struct {
+	podFullName string
+	status      api.PodStatus
+}
+
+type statusManager struct {
+	kubeClient client.Interface
+	// Map from pod full name to sync status of the corresponding pod.
+	podStatusesLock  sync.RWMutex
+	podStatuses      map[string]api.PodStatus
+	podStatusChannel chan podStatusSyncRequest
+}
+
+func newStatusManager(kubeClient client.Interface) *statusManager {
+	return &statusManager{
+		kubeClient:       kubeClient,
+		podStatuses:      make(map[string]api.PodStatus),
+		podStatusChannel: make(chan podStatusSyncRequest, 1000), // Buffer up to 1000 statuses
+	}
+}
+
+func (s *statusManager) Start() {
+	// We can run SyncBatch() often because it will block until we have some updates to send.
+	go util.Forever(s.SyncBatch, 0)
```

Comment:
What I don't like about the current approach is that it ~triples the runtime of the Kubelet tests (from ~2.5s to ~7.5s) since we're constantly waiting that 1s.